bibliopixel/control/api.py

```diff
@@ -0,0 +1,56 @@
+from http.server import HTTPServer, BaseHTTPRequestHandler
+import urllib
+import getpass, platform, sys, threading
+from .. util import log
+from . control import ExtractedControl
+
+class Api(ExtractedControl):
+    EXTRACTOR = {
+        'keys_by_type': {
+            'direct_command': ['action'],
+            'value_command': ['action', 'value']
+        }
+    }
+
+    def make_value_commands(self, value_command_list):
+        return [{'action': com, 'value': val, 'type': 'value_command'} for (com, val) in value_command_list]
+
+    def make_direct_commands(self, direct_command_list):
+        if direct_command_list == None:
+            return []
+        else:
+            return [{'action': com, 'type': 'direct_command'} for (com) in direct_command_list]
+
+    def make_messages(self, params):
+        parsed = (urllib.parse.parse_qs(params[2:]))
+        vcs = parsed.get('value_command')
+        vals = parsed.get('value')
+        if (vcs != None and vals != None):
+            vcl = list(zip(parsed.get('value_command'), parsed.get('value')))
+            value_commands = self.make_value_commands(vcl)
+        else:
+            value_commands = []
+        direct_commands = self.make_direct_commands(parsed.get('direct_command'))
+        messages = value_commands + direct_commands
+        for msg in messages:
+            self.receive(msg)
+
+    def _make_thread(self):
+        server_address = ('', 8080)
+        httpd = ApiServer(server_address, MsgRequestHandler)
+        api = self
+        thread = threading.Thread(target = httpd.serve_forever, args = (api, ))
+        return thread
+
+class ApiServer(HTTPServer):
+    def serve_forever(self, api):
+        self.RequestHandlerClass.api = api
```

Comment:
This seems a bit weird.  What's going on here?  

If `RequestHandlerClass` is a class (hardly unreasonable guess!) then an _instance_ is setting a static class variable.  That's a code smell - for example, if you create a second instance of `ApiServer`, even if you never start it, you might overwrite that variable.

Reply:
dunno, that was basically what I came up with for being able to pass an object into a request handler after googling for a few hours :D

Reply:
Innarestinge!  :-D

Before the review is done, I'll read the code and find out how it actually works.
My secret:

* `python`
* `import http.server`
*  `http.server.__file__`

and then edit that file.  But no need to do that unless you're bored, I'll get to it!

Reply:
no worries, i'll play with some refactoring this weekend for sure and I'll also get those new midi animations updated